transport/src/test/java/io/netty/channel/AutoFlushTest.java

```diff
@@ -0,0 +1,176 @@
+        environment.connect();
+        final ChannelPromise aggreggatedPromise = environment.clientChannel.newPromise();
+        ByteBuf data = newDataBuffer();
+        PromiseCombiner promiseCombiner = new PromiseCombiner();
+        for (int i = 0; i < 10; i++) {
+            ChannelPromise promise = environment.clientChannel.newPromise();
+            promiseCombiner.add(promise);
+            environment.clientChannel.write(data.retainedDuplicate(), promise);
+        }
+        promiseCombiner.finish(aggreggatedPromise);
+        aggreggatedPromise.sync();
+    }
+
+    @Test(timeout = 20000)
+    public void testFromWithinEventloop() throws Exception {
+        environment.connect();
+        final AtomicReference<ChannelFuture> writeResult = new AtomicReference<ChannelFuture>();
+        environment.clientChannel.eventLoop().submit(new Runnable() {
+            @Override
+            public void run() {
+                writeResult.set(environment.pipeline.write(newDataBuffer()));
+            }
+        }).sync();
+
+        writeResult.get().sync();
+    }
+
+    private static ByteBuf newDataBuffer() {
+        byte[] dataArr = new byte[32];
+        ThreadLocalRandom.current().nextBytes(dataArr);
+        return Unpooled.wrappedBuffer(dataArr);
+    }
+
+    public static final class TestEnvironment {
+
+        private final SingleThreadEventLoop serverEventloop;
+        private final SingleThreadEventLoop clientEventloop;
+        private final ServerBootstrap serverBootstrap;
+        private final Bootstrap bootstrap;
+        private ChannelPipeline pipeline;
+        private Channel clientChannel;
+        private Channel serverChannel;
+
+        private TestEnvironment(SingleThreadEventLoop serverEventloop, SingleThreadEventLoop clientEventloop,
+                                Class<? extends ServerChannel> serverChannelClass,
+                                Class<? extends Channel> clientChannelClass) {
+            this.serverEventloop = serverEventloop;
+            this.clientEventloop = clientEventloop;
+            serverBootstrap = new ServerBootstrap();
+            serverBootstrap.group(serverEventloop)
+                           .channel(serverChannelClass)
+                           .childHandler(new ChannelInitializer<Channel>() {
+                  @Override
+                  protected void initChannel(Channel ch) throws Exception {
+                      // Nothing to do.
+                  }
+              });
+            bootstrap = new Bootstrap();
+            bootstrap.option(ChannelOption.AUTO_FLUSH, true)
+                     .group(clientEventloop)
+                     .channel(clientChannelClass)
+                     .handler(new ChannelInitializer<Channel>() {
+                         @Override
+                         protected void initChannel(Channel ch) throws Exception {
+                             ch.pipeline().addFirst(new LoggingHandler());
```

Comment:
@NiteshKant please add a handler that calls release on all received messages so we not leak any buffers.